### PR TITLE
api: recomputeRoute field in JWT

### PR DESCRIPTION
**Pull request by arkodg**
Add a field called `useForRouting` that signals to Envoy Gateway that the headers generated from the claims are used to make routing decisions

Internally this field will be used to
~~* insert a catch-all route with a 404 direct response identical to https://github.com/envoyproxy/gateway/pull/2586 which makes sure the jwt filter with `claimToHeader` is applied before recomputing routing decision~~
* enable `clear_route_cache` to recompute routing decision https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/jwt_authn/v3/config.proto#extensions-filters-http-jwt-authn-v3-jwtprovider

Relates to https://github.com/envoyproxy/gateway/issues/2452

### CODE REVIEW

**Comment by arkodg** [comment]
outlining a few other alternatives
* add a similar field at top level within `jwt` instead of creating it within `jwt.claimToHeaders` , which will have the same effect internally, since these settings internally are applied at top level
* automatically enable these settings w/o an API field if `jwt.claimToHeaders` is set. One downside here is, enabling `clearRouteCache` is expensive and may impact performance. This may not be ideal for a user who wants to `jwt.claimToHeaders` for ratelimiting

File: api/v1alpha1/jwt_types.go [hunk 1]
```diff
@@ -85,6 +85,10 @@ type ClaimToHeader struct {
 	// (eg. "claim.nested.key", "sub"). The nested claim name must use dot "."
 	// to separate the JSON name path.
 	Claim string `json:"claim"`
+
+	// UseForRouting must be enabled if this header generated from the claim should be used for
+	// route matching decisions
+	UseForRouting *bool `json:"useForRouting,omitempty"`
```
Comment (shawnh2) [inline api/v1alpha1/jwt_types.go:91]: need a `// +optional` comment for this

Comment (arkodg) [reply api/v1alpha1/jwt_types.go:91]: thanks for catching this !

**Review by shawnh2 — APPROVED** [review]
lgtm

**Review by yuluo-yx — APPROVED** [review]
LGTM!

**Comment by arkodg** [comment]
ptal @guydc @zhaohuabing, I've limited the use of the field to only recalculate route, and not add a fallback/catch-all route, 
we can use docs to add that manually

**Comment by arkodg** [comment]
also looking at advice on API field naming, this is what ChatGPT said

```
Here are some alternate API field names for recomputeRoute:

calculateRoute
updateRoute
computeNewRoute
refreshRoute
recalculateRoute
```

File: api/v1alpha1/jwt_types.go [hunk 2]
```diff
@@ -52,8 +52,16 @@ type JWTProvider struct {
...
 	// The claim must be of type; string, int, double, bool. Array type claims are not supported
 	//
+	// +optional
 	ClaimToHeaders []ClaimToHeader `json:"claimToHeaders,omitempty"`
 
+	// RecomputeRoute clears the route cache and recalculates the routing decision.
+	// This field must be enabled if the headers generated from the claim are used for
+	// route matching decisions.
+	//
+	// +optional
```
Comment (guydc) [inline api/v1alpha1/jwt_types.go:62]: Do we need to state something about the impact on other features? e.g. filters that fired prior to jwt (CORS, ExtAuth, Basic, ...) will only execute in the context of the initial match?

**Review by guydc — COMMENTED** [review]
+1 to `updateRoute`

File: api/v1alpha1/jwt_types.go [hunk 3]
```diff
@@ -52,8 +52,16 @@ type JWTProvider struct {
...
+	RecomputeRoute *bool `json:"recomputeRoute,omitempty"`
```
Comment (zhaohuabing) [inline api/v1alpha1/jwt_types.go:63]: Should we verify that when recomputeRoute is enabled, the length of claimToHeaders is not zero?

**Comment by zhaohuabing** [comment]
recomputeRoute sounds good to me, "updateRoute" is a bit ambiguous.

File: api/v1alpha1/jwt_types.go (hunk 3 above)
Comment (arkodg) [reply api/v1alpha1/jwt_types.go:63]: yes !

**Review by zirain — APPROVED** [review]

File: api/v1alpha1/jwt_types.go (hunk 3 above)
Comment (zirain) [reply api/v1alpha1/jwt_types.go:63]: can you update cel validation?

File: api/v1alpha1/jwt_types.go (hunk 2 above)
Comment (arkodg) [reply api/v1alpha1/jwt_types.go:64]: added a comment around this

File: api/v1alpha1/jwt_types.go (hunk 3 above)
Comment (arkodg) [reply api/v1alpha1/jwt_types.go:65]: done ! thanks @zhaohuabing for helping me with my first CEL expression :)

**Review by zhaohuabing — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**